authorize_write.go

```diff
@@ -49,27 +40,44 @@ func (f *Fosite) WriteAuthorizeResponse(rw http.ResponseWriter, ar AuthorizeRequ
 		wh.Set(k, rh.Get(k))
 	}
 
-	// Implicit grants
-	// The endpoint URI MUST NOT include a fragment component.
-	redir.Fragment = ""
+	wh.Set("Cache-Control", "no-store")
+	wh.Set("Pragma", "no-cache")
+
+	redir := ar.GetRedirectURI()
 
-	u := redir.String()
+	switch ar.GetResponseMode() {
+	case ResponseModePost:
+		//form_post
+		rw.Header().Add("Content-Type", "text/html;charset=UTF-8")
+		WriteAuthorizeFormPostResponse(redir.String(), resp.GetParameters(), rw)
+	case ResponseModeQuery, ResponseModeNone:
+		// Explicit grants
+		q := redir.Query()
+		rq := resp.GetParameters()
+		for k := range rq {
+			q.Set(k, rq.Get(k))
+		}
+		redir.RawQuery = q.Encode()
+		sendRedirect(redir.String(), rw)
+	case ResponseModeFragment:
+		// Implicit grants
+		// The endpoint URI MUST NOT include a fragment component.
+		redir.Fragment = ""
 
-	fr := resp.GetFragment()
-	if len(fr) > 0 {
+		u := redir.String()
```

Comment:
Doing so double encode fragment. Once while doing `resp.GetParameters().Encode()` and then`redir.String()`

Reply:
Yes, this is ugly but there is no `RawFragment` (and newer versions of go where there is, it does not have the same semantics as `RawQuery`). So this has to be done like @ajanthan made and this is also how it was before.

I would suggest though that `fr.Encode()` + `plusMatch.ReplaceAllString(u, "%20")` is made into a helper function like `FormUrlEncode`. I am not sure though why we are replacing `+` with `%20`, because [`application/x-www-form-urlencoded`](https://www.w3.org/TR/1999/REC-html401-19991224/interact/forms.html#h-17.13.4.1) does use `+` for space (`application/x-www-form-urlencoded` is what fragment should be encoded with by OAuth spec).

Reply:
Done